README
WWW-Google-CustomSearch

The  README  is  used  to  introduce the module and provide instructions on how to install the
module, any machine dependencies it may have (for example C compilers and installed libraries)
and any other information that should be provided before the module is installed.

A  README  file is required for CPAN modules since CPAN extracts the README file from a module
distribution  so  that  people  browsing the archive can use it to get an idea of the module's
uses.  It is usually a good idea to provide version information here so that people can decide
whether fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

Alternatively, to install with Module::Build, you can use the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc WWW::Google::CustomSearch

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Google-CustomSearch

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/WWW-Google-CustomSearch

    CPAN Ratings
        http://cpanratings.perl.org/d/WWW-Google-CustomSearch

    Search CPAN
        http://search.cpan.org/dist/WWW-Google-CustomSearch/

LICENSE AND COPYRIGHT

Copyright (C) 2011 Mohammad S Anwar

This  program  is  distributed  in  the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

See http://dev.perl.org/licenses/ for more information.

Changes
Revision history for WWW-Google-CustomSearch

0.18    Fri Aug 29 10:10:00 2014
        - Added secction "REPOSITORY" to the pod
        - Added key "META_MERGE" to Makefile.PL
        - Updated Copyright year details.

0.17    Tue Jul 29 15:08:15 2014
        - Made the pre-requisite WWW::Google::UserAgent v0.07

0.16    Tue Jul 29 00:20:19 2014
        - Made the pre-requisite WWW::Google::UserAgent v0.06

0.15    Mon Jul 28 14:16:19 2014
        - Made the pre-requisite WWW::Google::UserAgent v0.03

0.14    Sun Jul 27 11:27:00 2014
        - Upgraded to use WWW::Google::UserAgent v0.02
        - Removed Build.PL

0.13    Sat Jul 19 04:00:00 2014
        - Restructured code to use Moo.
        - Making use of common library WWW::Google::UserAgent

0.12    Sat Mar 29 08:27:13 2014
        - Added dependency check for LWP::Protocol::https v6.04 (RT: 94194).

0.11    Sat Mar 01 11:12:13 2014
        - Added link to the Google JSON/Atom Custom Search API in the pod.

0.10    Sun Feb 02 17:15:10 2014
        - Fixed the issue raised by JACINTA (RT: 92596)

0.09    Fri Sep 14 08:55:10 2012
        - Fixed the warning when no items found.
        - Explicitly set the defaults.
        - Hide the 'alt' attribute from user as it is no longer relevant.

0.08    Fri Sep 14 06:02:22 2012
        - Added XML::Simple dependency check.

0.07    Thu Sep 13 15:00:00 2012
        - Simplied the process of fetching search result.
        - Fixed the handling of 'alt' attribute.

0.06    Wed May 02 11:15:00 2012
        - Updated pod about how the query parameters can be specified.

0.05    Tue Nov 15 11:12:13 2011
        - Fixed the issued raised by WILLERT (RT: 72417).

0.04    Mon Sep 12 17:18:19 2011
        - Fixed the checking of fields 'num' and 'start'.
        - Added unit test to keep an eye.

0.03    Mon Sep 12 14:15:16 2011
        - Handle default settings correctly.

0.02    Mon Aug 08 12:44:11 2011
        - Made Readonly v1.03 as minimum requirement.
        - Prepared the distribution using the Module::Signature v0.68.

0.01    Wed May 18 19:39:01 2011
        - First version, released.

Build.PL
use strict;use warnings;
use Module::Build;

my $builder = Module::Build->new(
    sign              => 1,
    module_name       => 'WWW::Google::CustomSearch',
    license           => 'perl',
    dist_author       => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    dist_version_from => 'lib/WWW/Google/CustomSearch.pm',
    add_to_cleanup    => [ 'WWW-Google-CustomSearch-*' ],
    build_requires    => {
        'Test::More'  => 0,
        'Readonly'    => '1.03',
        'Moose'       => '2.0006',
        'JSON'        => '2.15',
        'XML::Simple' => '2.18',
        'HTTP::Request'    => '1.40',
        'LWP::UserAgent'   => '2.33',
        'LWP::Protocol::https' => '6.04',
        'namespace::clean'     => '0.20',
        'MooseX::Params::Validate'     => '0.16',
        'Moose::Util::TypeConstraints' => '2.0006',
    },
);

$builder->create_build_script();

lib/WWW/Google/CustomSearch/Item.pm
package WWW::Google::CustomSearch::Item;

use Moose;
use namespace::clean;

use Data::Dumper;

=head1 NAME

WWW::Google::CustomSearch::Item - Placeholder for Google JSON/Atom Custom Search Item.

=head1 VERSION

Version 0.12

=cut

our $VERSION = '0.12';

has 'kind'             => (is => 'ro', isa => 'Str');
has 'link'             => (is => 'ro', isa => 'Str');
has 'displayLink'      => (is => 'ro', isa => 'Str');
has 'snippet'          => (is => 'ro', isa => 'Str');
has 'htmlSnippet'      => (is => 'ro', isa => 'Str');
has 'cacheId'          => (is => 'ro', isa => 'Str');
has 'formattedUrl'     => (is => 'ro', isa => 'Str');
has 'htmlFormattedUrl' => (is => 'ro', isa => 'Str');
has 'title'            => (is => 'ro', isa => 'Str');
has 'htmlTitle'        => (is => 'ro', isa => 'Str');

=head1 DESCRIPTION

=head1 METHODS

=head2 kind()

Returns the 'kind' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item ($result->items) {
        print "Item Kind: ", $item->kind, "\n" if defined $item->kind;
    }

=head2 link()

Returns the 'link' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item ($result->items) {
        print "Item Link: ", $item->link, "\n" if defined $item->link;
    }

=head2 displayLink()

Returns the 'displayLink' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item ($result->items) {
        print "Item Display Link: ", $item->displayLink, "\n" if defined $item->displayLink;
    }

=head2 snippet()

Returns the 'snippet' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item ($result->items) {
        print "Item Snippet: ", $item->snippet, "\n" if defined $item->snippet;
    }

=head2 htmlSnippet()

Returns the 'htmlSnippet' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item ($result->items) {
        print "Item HTML Snippet: ", $item->htmlSnippet, "\n" if defined $item->htmlSnippet;
    }

=head2 cacheId()

Returns the 'cacheId' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item ($result->items) {
        print "Item Cache Id: ", $item->cacheId, "\n" if defined $item->cacheId;
    }

=head2 formattedUrl()

Returns the 'formattedUrl' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item ($result->items) {
        print "Item Formatted URL: ", $item->formattedUrl, "\n" if defined $item->formattedUrl;
    }

=head2 htmlFormattedUrl()

Returns the 'htmlFormattedUrl' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item ($result->items) {
        print "Item HTML Formatted URL: ", $item->htmlFormattedUrl, "\n" if defined $item->htmlFormattedUrl;
    }

=head2 title()

Returns the 'title' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item ($result->items) {
        print "Item Title: ", $item->title, "\n" if defined $item->title;
    }

=head2 htmlTitle()

Returns the 'htmlTitle' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item ($result->items) {
        print "Item HTML Title: ", $item->htmlTitle, "\n" if defined $item->htmlTitle;
    }

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 BUGS

Please report  any  bugs or feature requests to C<bug-www-google-customsearch at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Google-CustomSearch>.
I will be notified and then you'll automatically be notified of progress on your bug as I make
changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Google::CustomSearch::Item

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Google-CustomSearch>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Google-CustomSearch>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Google-CustomSearch>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Google-CustomSearch/>

=back

=head1 LICENSE AND COPYRIGHT

This  program  is  free  software; you can redistribute it and/or modify it under the terms of
either:  the  GNU  General Public License as published by the Free Software Foundation; or the
Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 DISCLAIMER

This  program  is  distributed  in  the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut

__PACKAGE__->meta->make_immutable;
no Moose; # Keywords are removed from the WWW::Google::CustomSearch::Item package

1; # End of WWW::Google::CustomSearch::Item

lib/WWW/Google/CustomSearch/Page.pm
package WWW::Google::CustomSearch::Page;

use Moose;
use namespace::clean;

use Data::Dumper;
use Moose::Util::TypeConstraints;
use WWW::Google::CustomSearch;

=head1 NAME

WWW::Google::CustomSearch::Page - Placeholder for Google JSON/Atom Custom Search Page.

=head1 VERSION

Version 0.12

=cut

our $VERSION = '0.12';

type 'SafetyLevel'    => where { /\bhigh\b|\bmedium\b|\boff\b/i };
has  'api_key'        => (is => 'ro', isa => 'Str', required => 1);
has  'cx'             => (is => 'ro', isa => 'Str', required => 1);
has  'safe'           => (is => 'ro', isa => 'SafetyLevel', default => 'off');
has  'count'          => (is => 'ro', isa => 'Num');
has  'searchTerms'    => (is => 'ro', isa => 'Str', required => 1);
has  'inputEncoding'  => (is => 'ro', isa => 'Str', required => 1);
has  'startIndex'     => (is => 'ro', isa => 'Num', default  => 1);
has  'title'          => (is => 'ro', isa => 'Str', required => 1);
has  'totalResults'   => (is => 'ro', isa => 'Num', required => 1);
has  'outputEncoding' => (is => 'ro', isa => 'Str', required => 1);

=head1 DESCRIPTION

Provides the interface to the individual search page based on the search criteria.

=head1 METHODS

=head2 safe()

Returns the safety level of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Safety Level: ", $page->safe, "\n";

=head2 count()

Returns the 'count' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Page count: ", $page->count, "\n";

=head2 searchTerms()

Returns the 'searchTerms' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Search Terms: ", $page->searchTerms, "\n";

=head2 startIndex()

Returns the 'startIndex' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Start Index: ", $page->startIndex, "\n";

=head2 title()

Returns the 'title' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Title: ", $page->title, "\n";

=head2 totalResults()

Returns the 'totalResults' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Total Results: ", $page->totalResults, "\n";

=head2 inputEncoding()

Returns the 'inputEncoding' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Input Encoding: ", $page->inputEncoding, "\n";

=head2 outputEncoding()

Returns the 'outputEncoding' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Output Encoding: ", $page->outputEncoding, "\n";

=head2 fetch()

Perform a fresh search based on the previous input data and returns L<WWW::Google::CustomSearch::Result>
object.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    my $next    = $page->fetch;

=cut

sub fetch {
    my $self   = shift;
    my $engine = WWW::Google::CustomSearch->new({ api_key => $self->api_key,
                                                  cx      => $self->cx,
                                                  start   => $self->startIndex });
    return $engine->search($self->searchTerms);
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 BUGS

Please report  any  bugs or feature requests to C<bug-www-google-customsearch at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Google-CustomSearch>.
I will be notified and then you'll automatically be notified of progress on your bug as I make
changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Google::CustomSearch::Page

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Google-CustomSearch>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Google-CustomSearch>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Google-CustomSearch>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Google-CustomSearch/>

=back

=head1 LICENSE AND COPYRIGHT

This  program  is  free  software; you can redistribute it and/or modify it under the terms of
either:  the  GNU  General Public License as published by the Free Software Foundation; or the
Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 DISCLAIMER

This  program  is  distributed  in  the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut

__PACKAGE__->meta->make_immutable;
no Moose; # Keywords are removed from the WWW::Google::CustomSearch::Page package

1; # End of WWW::Google::CustomSearch::Page

lib/WWW/Google/CustomSearch/Result.pm
package WWW::Google::CustomSearch::Result;

use Moose;
use namespace::clean;

use Data::Dumper;
use WWW::Google::CustomSearch::Request;
use WWW::Google::CustomSearch::Page;
use WWW::Google::CustomSearch::Item;

=head1 NAME

WWW::Google::CustomSearch::Result - Placeholder for Google JSON/Atom Custom Search Result.

=head1 VERSION

Version 0.12

=cut

our $VERSION = '0.12';

has 'api_key' => (is => 'ro', isa => 'Str',     required => 1);
has 'raw'     => (is => 'ro', isa => 'HashRef', required => 1);

has 'kind'                  => (is => 'ro', isa => 'Str');
has 'formattedTotalResults' => (is => 'ro', isa => 'Str');
has 'formattedSearchTime'   => (is => 'ro', isa => 'Str');
has 'totalResults'          => (is => 'ro', isa => 'Str');
has 'searchTime'            => (is => 'ro', isa => 'Str');
has 'url_template'          => (is => 'ro', isa => 'Str');
has 'url_type'              => (is => 'ro', isa => 'Str');
has 'request'               => (is => 'ro', isa => 'WWW::Google::CustomSearch::Request');
has 'nextPage'              => (is => 'ro', isa => 'WWW::Google::CustomSearch::Page');
has 'previousPage'          => (is => 'ro', isa => 'WWW::Google::CustomSearch::Page');
has '_items'                => (is => 'ro', isa => 'ArrayRef[WWW::Google::CustomSearch::Item]');

sub BUILD
{
    my $self = shift;
    my $raw  = $self->raw;

    $self->{'kind'} = $raw->{'kind'};
    $self->{'formattedTotalResults'} = $raw->{'searchInformation'}->{'formattedTotalResults'};
    $self->{'formattedSearchTime'} = $raw->{'searchInformation'}->{'formattedSearchTime'};
    $self->{'totalResults'} = $raw->{'searchInformation'}->{'totalResults'};
    $self->{'searchTime'} = $raw->{'searchInformation'}->{'searchTime'};

    $self->{'url_template'} = $raw->{'url'}->{'template'};
    $self->{'url_type'} = $raw->{'url'}->{'type'};

    $raw->{'queries'}->{'request'}->[0]->{'api_key'} = $self->api_key;
    $self->{'request'} = WWW::Google::CustomSearch::Request->new($raw->{'queries'}->{'request'}->[0]);

    if (defined $raw->{'queries'}->{'nextPage'} && (scalar(@{$raw->{'queries'}->{'nextPage'}}))) {
        $raw->{'queries'}->{'nextPage'}->[0]->{'api_key'} = $self->api_key;
        $self->{'nextPage'} = WWW::Google::CustomSearch::Page->new($raw->{'queries'}->{'nextPage'}->[0]);
    }

    if (defined $raw->{'queries'}->{'previousPage'} && (scalar(@{$raw->{'queries'}->{'previousPage'}}))) {
        $raw->{'queries'}->{'previousPage'}->[0]->{'api_key'} = $self->api_key;
        $self->{'previousPage'} = WWW::Google::CustomSearch::Page->new($raw->{'queries'}->{'previousPage'}->[0]);
    }

    foreach (@{$raw->{items}}) {
        push @{$self->{_items}}, WWW::Google::CustomSearch::Item->new($_);
    }
}

=head1 DESCRIPTION

Provides the interface to the individual search results based on the search criteria.

=head1 METHODS

=head2 kind()

Returns the 'kind' attribute of the search result.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    print "Kind: ", $result->kind, "\n";

=head2 formattedTotalResults()

Returns the 'formattedTotalResults' attribute of the search result.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    print "Formatted Total Results: ", $result->formattedTotalResults, "\n";

=head2 formattedSearchTime()

Returns the 'formattedSearchTime' attribute of the search result.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    print "Formatted Search Time: ", $result->formattedSearchTime, "\n";

=head2 totalResults()

Returns the 'totalResults' attribute of the search result.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    print "Total Results: ", $result->totalResults, "\n";

=head2 searchTime()

Returns the 'searchTime' attribute of the search result.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    print "Search Time: ", $result->searchTime, "\n";

=head2 url_template()

Returns the URL template attribute of the search result.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    print "URL Template: ", $result->url_template, "\n";

=head2 url_type()

Returns the URL Type attribute of the search result.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    print "URL Type: ", $result->url_type, "\n";

=head2 request()

Returns the request L<WWW::Google::CustomSearch::Request> object used in the last search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $request = $result->request;

=head2 nextPage()

Returns the next page L<WWW::Google::CustomSearch::Page> object which can be used to fetch the
next page result.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;

=head2 previousPage()

Returns the previous page L<WWW::Google::CustomSearch::Page> object which can be used to fetch
the previous page result.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx, start => 2);
    my $result  = $engine->search("Google");
    my $page    = $result->previousPage;

=head2 items()

Returns list of search item L<WWW::Google::CustomSearch::Item> based on the search criteria.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $items   = $result->items; # ArrayRef
    my @items   = $result->items; # Array

=cut

sub items {
    my $self = shift;
    return unless defined $self->{_items};
    return @{$self->{_items}} if wantarray;
    return $self->{_items};
}

=head1 AUTHOR


Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 BUGS

Please report  any  bugs or feature requests to C<bug-www-google-customsearch at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Google-CustomSearch>.
I will be notified and then you'll automatically be notified of progress on your bug as I make
changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Google::CustomSearch::Result

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Google-CustomSearch>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Google-CustomSearch>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Google-CustomSearch>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Google-CustomSearch/>

=back

=head1 LICENSE AND COPYRIGHT

This  program  is  free  software; you can redistribute it and/or modify it under the terms of
either:  the  GNU  General Public License as published by the Free Software Foundation; or the
Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 DISCLAIMER

This  program  is  distributed  in  the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut

__PACKAGE__->meta->make_immutable;
no Moose; # Keywords are removed from the WWW::Google::CustomSearch::Result package

1; # End of WWW::Google::CustomSearch::Result

lib/WWW/Google/CustomSearch/Request.pm
package WWW::Google::CustomSearch::Request;

use Moose;
use namespace::clean;

use Data::Dumper;
use WWW::Google::CustomSearch::Page;

=head1 NAME

WWW::Google::CustomSearch::Request - Placeholder for Google JSON/Atom Custom Search Request.

=head1 VERSION

Version 0.12

=cut

our $VERSION = '0.12';

has 'page'    => (is => 'ro', isa => 'WWW::Google::CustomSearch::Page');
has 'api_key' => (is => 'ro', isa => 'Str', required => 1);

sub BUILD
{
    my $self = shift;
    my $args = shift;
    $args->{'api_key'} = $self->api_key;
    $self->{page} = WWW::Google::CustomSearch::Page->new($args);
}

=head1 DESCRIPTION

Provides the interface to the search request used last time.

=head1 METHOD

=head2 fetch()

Returns the L<WWW::Google::CustomSearch::Result> object base on the criteria used in the  last
search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->request;
    my $another = $page->fetch;

=cut

sub fetch {
    my $self = shift;
    return $self->{page}->fetch;
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 BUGS

Please report  any  bugs or feature requests to C<bug-www-google-customsearch at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Google-CustomSearch>.
I will be notified and then you'll automatically be notified of progress on your bug as I make
changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Google::CustomSearch::Request

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Google-CustomSearch>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Google-CustomSearch>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Google-CustomSearch>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Google-CustomSearch/>

=back

=head1 LICENSE AND COPYRIGHT

This  program  is  free  software; you can redistribute it and/or modify it under the terms of
either:  the  GNU  General Public License as published by the Free Software Foundation; or the
Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 DISCLAIMER

This  program  is  distributed  in  the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut

__PACKAGE__->meta->make_immutable;
no Moose; # Keywords are removed from the WWW::Google::CustomSearch::Request package

1; # End of WWW::Google::CustomSearch::Request

Makefile.PL
use strict;use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME          => 'WWW::Google::CustomSearch',
    AUTHOR        => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM  => 'lib/WWW/Google/CustomSearch.pm',
    ABSTRACT_FROM => 'lib/WWW/Google/CustomSearch.pm',
    dist          => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean         => { FILES => 'WWW-Google-CustomSearch-*' },
    SIGN          => 1,
    PREREQ_PM     => {
        'Test::More'  => 0,
        'Readonly'    => '1.03',
        'Moose'       => '2.0006',
        'JSON'        => '2.15',
        'XML::Simple' => '2.18',
        'HTTP::Request'    => '1.40',
        'LWP::UserAgent'   => '2.33',
        'LWP::Protocol::https' => '6.04',
        'namespace::clean'     => '0.20',
        'MooseX::Params::Validate'     => '0.16',
        'Moose::Util::TypeConstraints' => '2.0006',
    },
    ($ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
);
